File: pengcheng-core/pengcheng-realty/src/main/java/com/pengcheng/realty/unit/entity/RealtyUnit.java
package com.pengcheng.realty.unit.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 房源实体（一栋楼一层一户的颗粒度）
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@TableName("realty_unit")
public class RealtyUnit implements Serializable {

    // ---- 状态常量 ----
    public static final String STATUS_AVAILABLE   = "AVAILABLE";
    public static final String STATUS_RESERVED    = "RESERVED";
    public static final String STATUS_SUBSCRIBED  = "SUBSCRIBED";
    public static final String STATUS_SIGNED      = "SIGNED";
    public static final String STATUS_SOLD        = "SOLD";
    public static final String STATUS_UNAVAILABLE = "UNAVAILABLE";

    @TableId(type = IdType.AUTO)
    private Long id;

    /** 楼盘 ID */
    private Long projectId;

    /** 户型 ID */
    private Long houseTypeId;

    /** 楼栋号 */
    private String building;

    /** 楼层 */
    private Integer floor;

    /** 房号，如 0301 */
    private String unitNo;

    /** 完整编号，如 1-3-0301（building-floor-unitNo） */
    private String fullNo;

    /** 建筑面积 m² */
    private BigDecimal area;

    /** 挂牌价 */
    private BigDecimal listPrice;

    /** 实际成交价 */
    private BigDecimal actualPrice;

    /** 状态：AVAILABLE/RESERVED/SUBSCRIBED/SIGNED/SOLD/UNAVAILABLE */
    private String status;

    /** 锁定人 */
    private Long lockedBy;

    /** 锁定到期时间 */
    private LocalDateTime lockedUntil;

    /** 当前关联客户 */
    private Long customerId;

    /** 关联成交单 */
    private Long dealId;

    /** 备注 */
    private String remark;

    private LocalDateTime createTime;

    private LocalDateTime updateTime;
}

File: pengcheng-ui/src/api/realtyUnit.ts
import { request } from '@/utils/request'

// ---------- 类型定义 ----------

export interface HouseType {
  id?: number
  projectId: number
  code: string
  name: string
  bedrooms?: number
  livingRooms?: number
  bathrooms?: number
  area: number
  insideArea?: number
  orientation?: string
  layoutImage?: string
  basePrice?: number
  description?: string
  enabled?: number
  createTime?: string
}

export interface RealtyUnit {
  id?: number
  projectId: number
  houseTypeId: number
  building: string
  floor: number
  unitNo: string
  fullNo?: string
  area: number
  listPrice: number
  actualPrice?: number
  status?: string
  lockedBy?: number
  lockedUntil?: string
  customerId?: number
  dealId?: number
  remark?: string
  createTime?: string
}

export type UnitStatus = 'AVAILABLE' | 'RESERVED' | 'SUBSCRIBED' | 'SIGNED' | 'SOLD' | 'UNAVAILABLE'

export interface FloorRow {
  floor: number
  units: RealtyUnit[]
}

export interface UnitMatrix {
  building: string
  floors: FloorRow[]
}

export interface StatusChangeRequest {
  toStatus: UnitStatus
  operatorId?: number
  customerId?: number
  dealId?: number
  reason?: string
}

export interface LockRequest {
  userId: number
  hours?: number
}

// ---------- 户型 API ----------

export const houseTypeApi = {
  /** 按楼盘查询所有户型 */
  listByProject: (projectId: number) =>
    request.get<HouseType[]>('/admin/realty/house-types/by-project', { params: { projectId } }),

  /** 按楼盘查询启用户型 */
  listEnabled: (projectId: number) =>
    request.get<HouseType[]>('/admin/realty/house-types/enabled', { params: { projectId } }),

  /** 获取户型详情 */
  detail: (id: number) =>
    request.get<HouseType>(`/admin/realty/house-types/${id}`),

  /** 创建户型 */
  create: (data: HouseType) =>
    request.post<number>('/admin/realty/house-types', data),

  /** 编辑户型 */
  update: (id: number, data: HouseType) =>
    request.put<void>(`/admin/realty/house-types/${id}`, data),

  /** 删除户型 */
  delete: (id: number) =>
    request.delete<void>(`/admin/realty/house-types/${id}`),
}

// ---------- 房源 API ----------

export const unitApi = {
  /** 房源状态矩阵（楼栋 × 楼层 × 房间） */
  matrix: (projectId: number) =>
    request.get<UnitMatrix[]>('/admin/realty/units/matrix', { params: { projectId } }),

  /** 按状态筛选房源 */
  byStatus: (projectId: number, status?: UnitStatus) =>
    request.get<RealtyUnit[]>('/admin/realty/units/by-status', { params: { projectId, status } }),

  /** 获取房源详情 */
  detail: (id: number) =>
    request.get<RealtyUnit>(`/admin/realty/units/${id}`),

  /** 创建房源 */
  create: (data: RealtyUnit) =>
    request.post<number>('/admin/realty/units', data),

  /** 编辑房源 */
  update: (id: number, data: RealtyUnit) =>
    request.put<void>(`/admin/realty/units/${id}`, data),

  /** 删除房源 */
  delete: (id: number) =>
    request.delete<void>(`/admin/realty/units/${id}`),

  /** 变更房源状态 */
  changeStatus: (id: number, req: StatusChangeRequest) =>
    request.post<void>(`/admin/realty/units/${id}/change-status`, req),

  /** 原子锁定（返回 true 成功） */
  lock: (id: number, req: LockRequest) =>
    request.post<boolean>(`/admin/realty/units/${id}/lock`, req),

  /** 解锁 */
  unlock: (id: number) =>
    request.post<void>(`/admin/realty/units/${id}/unlock`),
}
